### PR TITLE
fix address alignment, required by cran

### PR DESCRIPTION
**Pull request by guolinke**

### CODE REVIEW

File: src/io/file_io.cpp [hunk 1]
```diff
@@ -47,6 +47,8 @@ struct LocalFile : VirtualFileReader, VirtualFileWriter {
   }
 
   size_t Write(const void* buffer, size_t bytes) const {
+    // 4 bytes alignment
+    CHECK_EQ(bytes % 4, 0);
```
Comment (jameslamb) [inline src/io/file_io.cpp:51]: it looks like this check is failing for many of the CI jobs (and not just R)

image

**Comment by jameslamb** [comment]
Ok I tried this using the steps in https://github.com/microsoft/LightGBM/pull/3338#issuecomment-700399862, got a similar error too the one showing up in CI

```text
[LightGBM] [Info] Saving data to binary file /tmp/RtmptKlgj2/lgb.Dataset_b6968fa090e
── 1. Error: lgb.Dataset: basic construction, saving, loading (@test_dataset.R#1
[LightGBM] [Fatal] Check failed: (bytes % 4) == (0) at io/file_io.cpp, line 51 .

Backtrace:
 1. lightgbm::lgb.Dataset.save(dtest1, tmp_file)
 2. dataset$save_binary(fname)
 3. lightgbm:::lgb.call(...)

[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000363 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
```

I have an early morning tomorrow so I'm going to get some sleep, but if you're able to fix it I can test this again tomorrow!

**Comment by guolinke** [comment]
@jameslamb It turns out that the fix is not trivial. many parts need to rewrite.

File: include/LightGBM/utils/file_io.h [hunk 2]
```diff
@@ -28,9 +29,20 @@ struct VirtualFileWriter {
    * \brief Append buffer to file
    * \param data Buffer to write from
    * \param bytes Number of bytes to write from buffer
+   * \param pad_to_4bytes
```
Comment (guolinke) [inline include/LightGBM/utils/file_io.h:32]:
```suggestion
```

File: include/LightGBM/utils/file_io.h [hunk 3]
```diff
@@ -28,9 +29,20 @@ struct VirtualFileWriter {
...
    * \return Number of bytes written
    */
   virtual size_t Write(const void* data, size_t bytes) const = 0;
+
+  size_t AlignedWrite(const void* data, size_t bytes) const {
+    auto ret = Write(data, bytes);
+    if (bytes % 4 != 0) {
+      size_t padding = bytes / 4 * 4 + 4 - bytes;
```
Comment (guolinke) [inline include/LightGBM/utils/file_io.h:40]:
```suggestion
      size_t padding = AlignedSize(bytes) - bytes;
```

**Comment by jameslamb** [comment]
> @jameslamb It turns out that the fix is not trivial. many parts need to rewrite.

thanks very much! I'll test this right now with the steps from https://github.com/microsoft/LightGBM/pull/3338#issuecomment-700399862

**Comment by jameslamb** [comment]
Ok I can see that the tests are all "passing" now, but I still see misalignment errors.

```text
Loading required package: R6
io/dataset_loader.cpp:373:76: runtime error: reference binding to misaligned address 0x55c61fb78f74 for type 'const value_type', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/bits/stl_vector.h:1198:21: runtime error: reference binding to misaligned address 0x55c61fb78f74 for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/bits/vector.tcc:449:28: runtime error: reference binding to misaligned address 0x55c61fb78f74 for type 'const type', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/bits/move.h:77:36: runtime error: reference binding to misaligned address 0x55c61fb78f74 for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/bits/alloc_traits.h:512:17: runtime error: reference binding to misaligned address 0x55c61fb78f74 for type 'const type', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/bits/move.h:77:36: runtime error: reference binding to misaligned address 0x55c61fb78f74 for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/ext/new_allocator.h:150:46: runtime error: reference binding to misaligned address 0x55c61fb78f74 for type 'const type', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/bits/move.h:77:36: runtime error: reference binding to misaligned address 0x55c61fb78f74 for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/ext/new_allocator.h:150:4: runtime error: load of misaligned address 0x55c61fb78f74 for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f74: note: pointer points here
  01 00 00 00 00 00 00 00  00 00 00 00 04 00 00 00  00 00 00 00 08 00 00 00  00 00 00 00 0b 00 00 00
              ^ 
/usr/include/c++/10/bits/stl_vector.h:1192:30: runtime error: reference binding to misaligned address 0x55c61fb78f8c for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f8c: note: pointer points here
  00 00 00 00 0b 00 00 00  00 00 00 00 0f 00 00 00  00 00 00 00 11 00 00 00  00 00 00 00 15 00 00 00
              ^ 
/usr/include/c++/10/bits/alloc_traits.h:512:17: runtime error: reference binding to misaligned address 0x55c61fb78f8c for type 'const type', which requires 8 byte alignment
0x55c61fb78f8c: note: pointer points here
  00 00 00 00 0b 00 00 00  00 00 00 00 0f 00 00 00  00 00 00 00 11 00 00 00  00 00 00 00 15 00 00 00
              ^ 
/usr/include/c++/10/bits/move.h:77:36: runtime error: reference binding to misaligned address 0x55c61fb78f8c for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f8c: note: pointer points here
  00 00 00 00 0b 00 00 00  00 00 00 00 0f 00 00 00  00 00 00 00 11 00 00 00  00 00 00 00 15 00 00 00
              ^ 
/usr/include/c++/10/ext/new_allocator.h:150:46: runtime error: reference binding to misaligned address 0x55c61fb78f8c for type 'const type', which requires 8 byte alignment
0x55c61fb78f8c: note: pointer points here
  00 00 00 00 0b 00 00 00  00 00 00 00 0f 00 00 00  00 00 00 00 11 00 00 00  00 00 00 00 15 00 00 00
              ^ 
/usr/include/c++/10/bits/move.h:77:36: runtime error: reference binding to misaligned address 0x55c61fb78f8c for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f8c: note: pointer points here
  00 00 00 00 0b 00 00 00  00 00 00 00 0f 00 00 00  00 00 00 00 11 00 00 00  00 00 00 00 15 00 00 00
              ^ 
/usr/include/c++/10/ext/new_allocator.h:150:4: runtime error: load of misaligned address 0x55c61fb78f8c for type 'const long unsigned int', which requires 8 byte alignment
0x55c61fb78f8c: note: pointer points here
  00 00 00 00 0b 00 00 00  00 00 00 00 0f 00 00 00  00 00 00 00 11 00 00 00  00 00 00 00 15 00 00 00
              ^ 

```

**Comment by guolinke** [comment]
Okay, they need 8 bytes alignment.

File: src/io/bin.cpp [hunk 4]
```diff
@@ -523,10 +523,11 @@ namespace LightGBM {
...
-    size += sizeof(bool);
+    size +=
+        static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(MissingType)));
+    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(bool)));
     size += sizeof(double);
-    size += sizeof(BinType);
+    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(BinType)));
     size += 2 * sizeof(double);
     size += bin * sizeof(double);
     size += sizeof(uint32_t) * 2;
```
Comment (guolinke) [inline src/io/bin.cpp:533]:
```suggestion
    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(uint32_t)) * 2;
```

File: src/io/bin.cpp [hunk 5]
```diff
@@ -523,10 +523,11 @@ namespace LightGBM {
   int BinMapper::SizeForSpecificBin(int bin) {
     int size = 0;
     size += sizeof(int);
```
Comment (guolinke) [inline src/io/bin.cpp:525]:
```suggestion
    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(int));
```

**Comment by guolinke** [comment]
@jameslamb now it should work

**Comment by jameslamb** [comment]
I see this in the CI jobs

```
/home/runner/work/LightGBM/LightGBM/lightgbm.Rcheck/00_pkg_src/lightgbm/src/src/io/bin.cpp: In static member function ‘static int LightGBM::BinMapper::SizeForSpecificBin(int)’:
/home/runner/work/LightGBM/LightGBM/lightgbm.Rcheck/00_pkg_src/lightgbm/src/src/io/bin.cpp:525:73: error: expected ‘)’ before ‘;’ token
     size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(int));
                                                                         ^
/home/runner/work/LightGBM/LightGBM/lightgbm.Rcheck/00_pkg_src/lightgbm/src/src/io/bin.cpp:533:82: error: expected ‘)’ before ‘;’ token
     size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(uint32_t)) * 2;
                                                                                  ^
CMakeFiles/_lightgbm.dir/build.make:157: recipe for target 'CMakeFiles/_lightgbm.dir/src/io/bin.cpp.o' failed
make[3]: *** [CMakeFiles/_lightgbm.dir/src/io/bin.cpp.o] Error 1
```

File: src/io/bin.cpp [hunk 6]
```diff
@@ -522,36 +522,37 @@ namespace LightGBM {
 
   int BinMapper::SizeForSpecificBin(int bin) {
     int size = 0;
-    size += sizeof(int);
-    size += sizeof(MissingType);
-    size += sizeof(bool);
+    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(int));
```
Comment (guolinke) [inline src/io/bin.cpp:525]:
```suggestion
    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(int)));
```

File: src/io/bin.cpp [hunk 7]
```diff
@@ -522,36 +522,37 @@ namespace LightGBM {
...
+    size +=
+        static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(MissingType)));
+    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(bool)));
     size += sizeof(double);
-    size += sizeof(BinType);
+    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(BinType)));
     size += 2 * sizeof(double);
     size += bin * sizeof(double);
-    size += sizeof(uint32_t) * 2;
+    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(uint32_t)) * 2;
```
Comment (guolinke) [inline src/io/bin.cpp:533]:
```suggestion
    size += static_cast<int>(VirtualFileWriter::AlignedSize(sizeof(uint32_t))) * 2;
```

**Comment by guolinke** [comment]
sorry, it is a typo caused by editing on the web page.

**Comment by jameslamb** [comment]
no problem! testing right now

**Comment by jameslamb** [comment]
@guolinke THIS WORKED!!  🎉 

No test failures, and no mis-alignment errors

**Comment by guolinke** [comment]
great! let us move forward to the memory leakage problem.

**Comment by jameslamb** [comment]
> great! let us move forward to the memory leakage problem.

ok sounds good!  I will approve this so we can merge it, and I'll cherry-pick it onto #3338 

I'm working on trying to reproduce the CRAN valgrind checks right now.

**Review by jameslamb — APPROVED** [review]

**Pull request merged**